### ISSUE TITLE
Infinite run or optimizer fail

### ISSUE THREAD

**Issue by dy-octa** [tensorflow/tensorflow#23159]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.9.0
- Python version: 3.6.0
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version:9.1
- GPU model and memory: P100 16GB

**Describe the current behavior**
sess.run() executes infinitely and will never return
In the alternative version, sess.run() produces expected results but error is produced by the optimizer (see log).

The bug only occurs when 
1. On GPU. The code executes properly on CPU.
2. The condition of tf.cond holds False, i.e. when tf.gradients is not to be executed. Otherwise correct gradients could be calculated.
3. There're loop structures, e.g. RNN.  

**Code to reproduce the issue**

> X = tf.placeholder(shape=[None, None, 10], dtype=tf.float32)
> Y, _ = tf.nn.bidirectional_dynamic_rnn(LSTMCell(20), LSTMCell(20), X, dtype=tf.float32)
> 
> loss = tf.reduce_mean(Y)
> vars = tf.trainable_variables()
> grads = tf.cond(tf.shape(X)[0] < 3, lambda :tf.gradients(loss, vars), lambda :[np.nan] * len(vars))
> 
> sess = tf.Session()
> sess.run(tf.global_variables_initializer())
> print(sess.run(grads, feed_dict={X: np.random.random([30, 10, 10])}))

Alternative version:

> X = tf.constant(np.random.random([30, 10, 10]), dtype=tf.float32)
> Y, _ = tf.nn.bidirectional_dynamic_rnn(LSTMCell(20), LSTMCell(20), X, dtype=tf.float32)
> 
> loss = tf.reduce_mean(Y)
> vars = tf.trainable_variables()
> grads = tf.cond(tf.shape(X)[0] < 3, lambda :tf.gradients(loss, vars), lambda :[np.nan] * len(vars))
> 
> sess = tf.Session()
> sess.run(tf.global_variables_initializer())
> print(sess.run(grads))

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

> 2018-10-22 17:32:36.399614: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:581] Iteration = 0, topological sort failed with message: The graph couldn't be sorted in topological order.
> 2018-10-22 17:32:36.401378: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:581] Iteration = 1, topological sort failed with message: The graph couldn't be sorted in topological order.
>

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Exact command to reproduce

**Comment by wt-huang** [comment]
@dy-octa I run the above alternate version using latest TensorFlow version and didn't receive any error.
TensorFlow nightly works as well. There was no error produced by the optimizer.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?
